### PR TITLE
Fuse the DSA (V3.2, GLM-5.x) indexer Q/K paths into single kernels

### PR DESCRIPTION
**Pull request by b8zhong**
<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.io to discuss further. -->

Note: to completely revert the affects of this PR, we can set `SGLANG_DISABLE_DSA_INDEXER_FUSION=1`

## Changes

1. Fuse the key projection and the head-gate projection into a single matmul
   instead of two.

2. Fuse the query path — rope, optional Hadamard, fp8 quantization, and
   head-gate scaling — into one kernel instead of four separate ops.

3. Fuse the key path — normalization, rope, optional Hadamard, fp8 quantization,
   and the index cache write — into one kernel instead of four. The previous
   split path is kept as a fallback for the cases that can't use the fused store
   (non-CUDA / NPU, HIP).

4. Drop the Hadamard rotation for V3.2. It cancels out of the indexer scores, so
   removing it doesn't change results — it only affected quantization accuracy,
   and V3.2 now quantizes the un-rotated activations directly. It stays available
   for the paths that still want it.

5. Run the key and query paths concurrently. Previously the cache store sat on
   the critical path after the key work; folding the store into the key kernel
   lets the entire key side overlap the query side.

Note: for the fusion, it seems vLLM has adopt a similar approach: https://github.com/vllm-project/vllm/pull/38928

## Checkpoint / weight loading

The fused key+head-gate matmul is bf16, so the loader has to land the key
projection as bf16. How it gets there depends on the checkpoint format (the bf16
result is identical either way):

| Checkpoint | indexer key projection | loader behavior |
|---|---|---|
| deepseek-ai/DeepSeek-V3.2 (FP8) | block-fp8 + scale | dequantize to bf16 at load |
| zai-org/GLM-5-FP8 | block-fp8 + scale | dequantize to bf16 at load |
| nvidia/DeepSeek-V3.2-NVFP4 | already bf16 | straight copy |
| nvidia/GLM-5.1-NVFP4 | already bf16 | straight copy |

The NVFP4 checkpoint authors intentionally left the (small) indexer out of
low-bit quantization and pre-dequantized its key projection to bf16, verified it
matches the FP8 checkpoint dequantized exactly. The head-gate projection is bf16
in all of them.

## FP4:
Before:
<img width="1686" height="913" alt="Screenshot 2026-06-09 at 10 56 53 AM" src="https://github.com/user-attachments/assets/b6b91219-986f-40fb-b394-2e3e3644e51e" />

After:
<img width="1678" height="917" alt="Screenshot 2026-06-09 at 10 58 15 AM" src="https://github.com/user-attachments/assets/4d98d719-7003-47b2-83e5-037b5c4a363b" />
<img width="1562" height="413" alt="Screenshot 2026-06-09 at 1 17 45 PM" src="https://github.com/user-attachments/assets/abaf3cdc-261c-4db6-aace-3f0a8d1b2f25" />

## FP8:
Before:
<img width="1680" height="485" alt="Screenshot 2026-06-12 at 6 53 53 AM" src="https://github.com/user-attachments/assets/e8e945eb-6779-4c2c-b109-8c56a492b470" />

After:
<img width="1261" height="523" alt="Screenshot 2026-06-12 at 6 54 06 AM" src="https://github.com/user-attachments/assets/adf27054-c79e-4f60-9b89-6650ad5a98d4" />

<!-- Detail the changes made in this pull request. -->

## Accuracy Tests

GPQA:
```
nohup env CUDA_VISIBLE_DEVICES=4,5,6,7 python3 -m sglang.launch_server \
  --model-path nvidia/GLM-5.1-NVFP4 \
  --tp 4 \
  --ep 4 \
  --trust-remote-code \
  --reasoning-parser glm45 \
  --tool-call-parser glm47 \
  --port 30001 \
  --max-running-requests 256 > glm_server.log 2>&1 &
```

```
sglang serve \
  --model-path zai-org/GLM-5.2-FP8 \
  --tp 8 \
  --ep 8 \
  --trust-remote-code \
  --reasoning-parser glm45 \
  --tool-call-parser glm47 \
  --max-running-requests 256
```

```
nohup python3 -m sglang.test.run_eval --port 30020 --eval-name gpqa --num-examples 198 --max-tokens 128000 --repeat 8 --top-p 0.95 --temperature 1.0 --thinking-mode glm-45 > eval.log 2>&1 &
```

## FP4:

Before:
```
Repeat: 8, mean: 0.854
Scores: ['0.838', '0.848', '0.869', '0.838', '0.854', '0.854', '0.864', '0.869']
```

After:
```
Repeat: 8, mean: 0.859
Scores: ['0.843', '0.828', '0.874', '0.864', '0.864', '0.848', '0.884', '0.864']
```

## FP8:
Before:
```
Repeat: 8, mean: 0.856
Scores: ['0.869', '0.838', '0.889', '0.828', '0.838', '0.864', '0.864', '0.859']
```

After:
```
Repeat: 8, mean: 0.856
Scores: ['0.859', '0.859', '0.843', '0.843', '0.848', '0.859', '0.848', '0.889']
```

<!-- If this pull request affects model outputs (e.g., changes to the kernel or model forward code), provide accuracy test results. -->

## Speed Tests and Profiling

BS = 1: 97.72 -> 107.77 TPS with 0 context (about 10%), with above command on B300.

For B300 + GLM FP8 checkpoint and BS = 128, around 3212.05 -> 3418.63 TPS in decode

<!-- If this pull request impacts inference speed, provide benchmarking and profiling results. -->





































































































































































































































































































































































































































<!-- pr-states:start -->
---
### CI States

Latest PR Test (Base): <!-- slot:pr-test:start -->:white_check_mark: [Run #28271629599](https://github.com/sgl-project/sglang/actions/runs/28271629599)<!-- slot:pr-test:end -->
Latest PR Test (Extra): <!-- slot:pr-test-extra:start -->:white_check_mark: [Run #28302367849](https://github.com/sgl-project/sglang/actions/runs/28302367849)<!-- slot:pr-test-extra:end -->
<!-- pr-states:end -->

### CODE REVIEW

**Comment by mattteochen** [comment]
I may be missing something, but I'm having a doubt about mixed index-K cache representations after dropping Hadamard in the fused V3.2 path.

My understanding is that Hadamard is skipped even though the called kernel still has `Hadamard` in its name. `fused_q_indexer_rope_first_quant` calls `_jit_main_q_indexer_rope_first_quant_module`, which instantiates `FusedQIndexerRopeHadamardQuantKernel` with `make_cpp_args(dtype, is_arch_support_pdl(), True, False)`. In the C++ template those last two args are `kRopeFirst=true` and `kHadamard=false`, and the Hadamard block is behind `if constexpr (kHadamard)`, so it should be compiled out for this path.

However, the prefill fast path still seems to go through:
`skip_logits_computation -> _forward_cuda_k_only -> _get_k_bf16 -> rotate_activation(key) -> _store_index_k_cache`

So for short prefill / extend where `max_kv_len <= index_topk`, the cache may be populated with Hadamard-rotated K.

Is this path impossible in the intended GLM-5/V3.2 decode flow, or should `_forward_cuda_k_only` also use the no-Hadamard K path when DSA indexer fusion is enabled? I don’t think `test_dsv32_indexer.py` covers this mixed prefill/decode cache case, since it tests the standalone fused kernels against no-Hadamard references.

**Comment by b8zhong** [comment]
@mattteochen Thanks for catching this bug. Also fixed it in the CP case

**Comment by b8zhong** [comment]
Note: it seems /rerun-stage will combine the nightly tests into 1 run, and have a fixed timeout of 1 hour, even though its' sum may exceed 1 hour (normally) (so the failing tests are not actually a result of a bug) (and actually, each of these tests need 90 minutes, so it's normal for them to be failing). Might make a fork to quickly test

**Comment by yuho8818** [comment]
LGTM

**Comment by b8zhong** [comment]
Here are the **latest results** with BCG (since merging conflicts is a little risky):
BS = 1 decode, ISL 10k

## FP8 decode
Commands:
```
SGLANG_ENABLE_PCG_DSV2_DUAL_STREAM=1 SGLANG_DISABLE_DSA_INDEXER_FUSION=1 \
sglang serve \
    --model-path zai-org/GLM-5.2-FP8 \
    --tp 4 \
    --cuda-graph-backend-prefill breakable
```
Before:
Screenshot 2026-06-24 at 6 15 53 PM

After:
Screenshot 2026-06-24 at 6 15 06 PM

## FP8 prefill (BCG)
Before:
Screenshot 2026-06-24 at 6 21 56 PM

After:
Screenshot 2026-06-24 at 6 22 06 PM

## FP8 prefill (PCG)
Before:
Screenshot 2026-06-24 at 6 50 54 PM

After:
Screenshot 2026-06-24 at 7 01 44 PM


## NVFP4 decode
Before:
Screenshot 2026-06-24 at 7 35 06 PM


After:
Screenshot 2026-06-24 at 7 34 49 PM

## NVFP4 prefill (PCG)
Before:
Screenshot 2026-06-24 at 7 36 08 PM

After:
Screenshot 2026-06-24 at 7 36 24 PM

**Comment by mattteochen** [comment]
> SGLANG_DISABLE_DSA_INDEXER_FUSION

It looks sweet! So, can `SGLANG_ENABLE_PCG_DSV2_DUAL_STREAM=1` and `SGLANG_DISABLE_DSA_INDEXER_FUSION=0` coexist nicely? At least this is my impression from the profiles.

Update: `SGLANG_ENABLE_PCG_DSV2_DUAL_STREAM=1` shouldn't have any effect when ISL is 10k because BCG/PCG capture has a 2k token limit by default on this model, if I'm not wrong. I got confused by the dual stream in the "after" profile.
Does the dual stream in prefill in the "after" screenshot come from the fused kernel implementation?

**Comment by b8zhong** [comment]
@mattteochen Actually the dual stream is a bit of an accident, it should be turned off when num tokens > 1024. However, it seems to not make a great performance difference (at least it is much better than the original). What do you think about that (also do you think we need to reprofile under num tokens = 1024?)

**Comment by mattteochen** [comment]
> @mattteochen Actually the dual stream is a bit of an accident, it should be turned off when num tokens > 1024. However, it seems to not make a great performance difference (at least it is much better than the original). What do you think about that (also do you think we need to reprofile under num tokens = 1024?)

I think we're good because this PR works on the indexer while the dual stream flag targets MoE. They shouldn't collide.

And yes, I don't think having the overlap is crucial here looking at the profile you posted.

File: python/sglang/jit_kernel/dsv32/__init__.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+"""DeepSeek-V3.2 only."""
```
Comment (Fridge003) [inline python/sglang/jit_kernel/dsv32/__init__.py]: DSA only

File: python/sglang/jit_kernel/dsv32/utils.py [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# DeepSeek-V3.2 only.
+def make_name(name: str) -> str:
```
Comment (Fridge003) [inline python/sglang/jit_kernel/dsv32/utils.py]: Do we need this

File: python/sglang/jit_kernel/csrc/deepseek_v4/main_norm_rope.cuh [hunk 3]
```diff
@@ -425,11 +425,13 @@ struct FusedQIndexerRopeHadamardQuantParams {
...
   const float* __restrict__ freqs_cis;  // (max_pos, 64) fp32
   const void* __restrict__ positions;   // (B,) PosT
+  // Row stride for `weight` (caller passes the non-contiguous wk slice directly).
+  int64_t weight_stride_batch;
   uint32_t batch_size;
   uint32_t num_heads;
 };
 
-template <typename DType, typename PosT, bool kUsePDL>
+template <typename DType, typename PosT, bool kUsePDL, bool kRopeFirst = false, bool kHadamard = true>
```
Comment (Fridge003) [inline python/sglang/jit_kernel/csrc/deepseek_v4/main_norm_rope.cuh:434]: I'm afraid these changes will break dsv4. 
Can we create some new functions dedicated for dsa models

File: python/sglang/srt/layers/attention/dsa/dsa_indexer.py [hunk 4]
```diff
@@ -53,6 +53,7 @@
 
 global _use_multi_stream
 _is_cuda = is_cuda()
+_use_dsa_indexer_fusion = _is_cuda and not envs.SGLANG_DISABLE_DSA_INDEXER_FUSION.get()
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/dsa/dsa_indexer.py:56]: Is it possible to create a new class for the fused indexer and put it under a new file?
So we don't need to add the if-else branches here. The codes can be much cleaner

File: python/sglang/srt/lora/lora_manager.py [hunk 5]
```diff
@@ -605,6 +606,21 @@ def init_lora_shapes(
                 # Otherwise, infer target_modules from adapter configs.
                 self.target_modules.update(adapter_target_modules)
 
+        # Fusion folds wk + weights_proj into wk_weights_proj, so the modules
```
Comment (Fridge003) [inline python/sglang/srt/lora/lora_manager.py:609]: We can assert this error in `server_args.py`. (Like when `--enable-lora` is true, assert dsa indexer fusion is False)
No need to fix this file

File: test/registered/jit/test_dsv32_indexer_fusion.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
+"""Correctness tests for the DeepSeek-V3.2 DSA indexer fused kernels.
```
Comment (Fridge003) [inline test/registered/jit/test_dsv32_indexer_fusion.py:1]: There is already a test called `test/registered/kernels/test_dsa_indexer.py`

We can rename this file to `test_dsv32_indexer_fusion.py`

File: python/sglang/srt/models/deepseek_common/deepseek_weight_loader.py [hunk 7]
```diff
@@ -76,6 +76,52 @@ def _clone_if_runai_streamed_tensor(tensor: torch.Tensor) -> torch.Tensor:
...
+    SGLANG_DISABLE_DSA_INDEXER_FUSION set, where wk and weights_proj are
+    separate) so the caller falls through to per-tensor loading.
+    """
+    fused_name = name.rsplit(".indexer.", 1)[0] + ".indexer.wk_weights_proj.weight"
+    fused_param = params_dict.get(fused_name)
+    if fused_param is None or fused_param.dtype != torch.bfloat16:
+        return False
+
+    if ".indexer.weights_proj." in name:
+        w = _clone_if_runai_streamed_tensor(loaded_weight)
```
Comment (Fridge003) [inline python/sglang/srt/models/deepseek_common/deepseek_weight_loader.py:100]: Why adding this runai thing

File: python/sglang/srt/models/deepseek_common/deepseek_weight_loader.py [hunk 8]
```diff
@@ -207,6 +255,19 @@ def do_load_weights(
                 if "rotary_emb.inv_freq" in name:
                     continue
 
+                # CUDA fuses wk + weights_proj into one bf16 wk_weights_proj; the
+                # helper returns True once it has consumed the shard.
+                if (
+                    ".indexer.wk." in name or ".indexer.weights_proj." in name
+                ) and _load_fused_indexer_wk(
```
Comment (Fridge003) [inline python/sglang/srt/models/deepseek_common/deepseek_weight_loader.py:262]: Do we need to check the flag SGLANG_DISABLE_DSA_INDEXER_FUSION here

File: python/sglang/jit_kernel/csrc/deepseek_v4/main_norm_rope.cuh (hunk 3 above)
Comment (b8zhong) [reply python/sglang/jit_kernel/csrc/deepseek_v4/main_norm_rope.cuh:434]: These arguments are set to the default to what DSV4 expects. Is that OK

File: python/sglang/jit_kernel/dsv32/utils.py (hunk 2 above)
Comment (b8zhong) [reply python/sglang/jit_kernel/dsv32/utils.py]: Ok, deleting. It's a (weird) convention from DSV4 actually

File: python/sglang/jit_kernel/dsv32/__init__.py (hunk 1 above)
Comment (b8zhong) [reply python/sglang/jit_kernel/dsv32/__init__.py]: Ok. Renamed it other places too

File: python/sglang/srt/models/deepseek_common/deepseek_weight_loader.py (hunk 7 above)
Comment (b8zhong) [reply python/sglang/srt/models/deepseek_common/deepseek_weight_loader.py:100]: It's needed to make the runai streamer work. For the streaming buffer of runai lodaer (but its hack maybe outside of this PR)

File: python/sglang/srt/models/deepseek_common/deepseek_weight_loader.py (hunk 8 above)
Comment (b8zhong) [reply python/sglang/srt/models/deepseek_common/deepseek_weight_loader.py:262]: It's created by the indexer init already. So it's redundant (also confirmed the previous path not affected)

File: test/registered/jit/test_dsv32_indexer_fusion.py (hunk 6 above)
Comment (b8zhong) [reply test/registered/jit/test_dsv32_indexer_fusion.py:1]: REnamed

File: python/sglang/srt/lora/lora_manager.py (hunk 5 above)
Comment (b8zhong) [reply python/sglang/srt/lora/lora_manager.py:609]: Reverted and done that

File: python/sglang/srt/layers/attention/dsa/dsa_indexer.py (hunk 4 above)
Comment (Fridge003) [reply python/sglang/srt/layers/attention/dsa/dsa_indexer.py:56]: Left for the future

**Review by Fridge003 — APPROVED** [review]
